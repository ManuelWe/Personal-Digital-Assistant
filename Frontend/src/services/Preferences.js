import axios from 'axios';

const api = axios.create({
  baseURL: '/api/',
});

const functions = {
  getPreferences() {
    return api.get('preferences');
  },
  updatePreferences(preferences) {
    return api.patch('preferences', preferences);
  },
};

export default functions;
