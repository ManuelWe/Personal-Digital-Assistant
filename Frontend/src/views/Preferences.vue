<template>
  <div
    id="preferences"
    class="hero-body"
  >
    <b-loading
      :is-full-page="false"
      :active.sync="loading"
      :can-cancel="true"
    />
    <div
      v-if="!loading"
    >
      <h1 class="title">
        Preferences for {{ user.name }}
      </h1>

      <div class="field">
        <b-checkbox
          v-model="notificationsEnabled"
          @input="toggleNotifications"
        >
          Send notifications to this device
        </b-checkbox>
      </div>

      <div class="field">
        <label class="label has-text-white">Name</label>
        <div class="control">
          <input
            :value="user.name"
            class="input"
            type="text"
            placeholder="Name"
          >
        </div>
      </div>

      <div class="field">
        <label class="label has-text-white">Preparation Time in Minutes</label>
        <div class="control">
          <input
            :value="user.preferences.preparationTimeInMinutes"
            class="input"
            type="text"
            placeholder="Preparation Time in Minutes"
          >
        </div>
      </div>

      <div class="field">
        <label class="label has-text-white">Favourite Quote Category</label>
        <div class="control">
          <input
            :value="user.preferences.quoteCategory"
            class="input"
            type="text"
            placeholder="Favourite Quote Category"
          >
        </div>
      </div>

      <div class="field">
        <label class="label has-text-white">Main City</label>
        <div class="control">
          <input
            :value="user.preferences.weatherCity"
            class="input"
            type="text"
            placeholder="Main City"
          >
        </div>
      </div>

      <div class="field is-grouped is-grouped-right">
        <p class="control">
          <router-link
            class="button is-light"
            :to="{name: 'landingPage'}"
          >
            Cancel
          </router-link>
        </p>
        <p class="control">
          <a
            class="button is-success"
            @click="savePreferences()"
          >
            Save
          </a>
        </p>
      </div>
    </div>
  </div>
</template>

<script>
import UserService from '../services/User';

export default {
  data() {
    return {
      user: null,
      loading: true,
    };
  },

  created() {
    const utterance = new SpeechSynthesisUtterance('Edit your Preferences.');
    utterance.rate = 1.3;
    utterance.lang = 'en-US';
    speechSynthesis.speak(utterance);

    UserService.getUser().then((result) => {
      this.user = result.data.data;
      this.loading = false;
    });
  },

  methods: {
    savePreferences() {
      console.log('SAVE');

      this.$buefy.toast.open({
        message: 'Data theoretically saved! (Not yet implemented.)',
        duration: 3000,
        type: 'is-success',
      });

      const utterance = new SpeechSynthesisUtterance('Saved successfully.');
      utterance.rate = 1.3;
      utterance.lang = 'en-US';
      speechSynthesis.speak(utterance);
    },

    async toggleNotifications() {
      if (this.notificationsEnabled) {
        await this.enableNotifications();
      } else {
        await this.disableNotifications();
      }
    },

    async enableNotifications() {
      try {
        const notificationPermission = await Notification.requestPermission();
        if (notificationPermission !== 'granted') {
          this.$buefy.snackbar.open({
            message: 'Unfortunately, Gunter doesn\'t have permission to send you notifications on this device. Please check your browser settings.',
            type: 'is-danger',
          });
          this.notificationsEnabled = false;
          return;
        }

        const serviceWorkerRegistration = await navigator.serviceWorker.ready;
        const pushSubscription = await serviceWorkerRegistration.pushManager.subscribe({
          userVisibleOnly: true,
          applicationServerKey: 'BBauGh8G3IdDf28vFQD0-Nn-8wniZUsCjvRa0F0MbRUTmy0NDDGQCT-OD3M5k8c54DNsyw9-_SwibbBXxWYG_nk',
        });

        const response = await fetch('/api/notifications/enable', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(pushSubscription),
        });
        if (!response.ok) {
          throw await response.text();
        }
      } catch (err) {
        console.error(err);
        this.$buefy.snackbar.open({
          message: 'Unfortunately, Gunter couldn\'t enable notifications for you on this device. He doesn\'t know why and is truly sorry.',
          type: 'is-danger',
        });
        this.notificationsEnabled = false;
      }
    },

    async disableNotifications() {
      try {
        const response = await fetch('/api/notifications/disable', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
        });
        if (!response.ok) {
          throw await response.text();
        }
      } catch (err) {
        console.error(err);
        this.$buefy.snackbar.open({
          message: 'Unfortunately, Gunter couldn\'t disable notifications for you on this device. He doesn\'t know why and is truly sorry.',
          type: 'is-danger',
        });
        this.notificationsEnabled = true;
      }
    },
  },
};
</script>

<style scoped>
.hero-body {
  position: relative;
  margin: 0rem 3rem 3rem 3rem;
}
</style>
