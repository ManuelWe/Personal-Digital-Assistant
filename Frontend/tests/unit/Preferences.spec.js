import { createLocalVue, mount } from '@vue/test-utils';
import Buefy from 'buefy';
import VueRouter from 'vue-router';
import mockAxios from 'axios';
import Preferences from '@/views/Preferences.vue';

const localVue = createLocalVue();
localVue.use(Buefy);
localVue.use(VueRouter);

const routes = [
  {
    path: '/',
    name: 'landingPage',
    component: Preferences,
  },
];
const router = new VueRouter({
  routes,
});

const mockPreferences = {
  data: {
    data: {
      name: 'Gunter',
      age: 107,
      favouriteFood: 'Jelly Beans',
      loyal: true,
      preferences: {
        calendarUrl: 'https://calendar.google.com/calendar/ical/rroff00labrg6qt5gu3ol87ejo%40group.calendar.google.com/private-97ee3f8105c50679da64c381b7890270/basic.ics',
        preparationTimeInMinutes: '45',
        quoteCategory: 'funny',
        weatherCity: 'Stuttgart',
        currentLocationCoordinates: '48.6616037,9.3501336',
        currentLocationAddress: 'Stuttgart, Schmidenerstraße 21',
        eventLocationCoordinates: '48.773563,9.171063',
        eventLocationAddress: 'Stuttgart, Rotebühlplatz 41',
      },
      events: [{ title: 'Party hard', date: '2020-03-05 14:10.00' }],
    },
  },
};

jest.mock('@/services/SpeechSynthesis');
jest.mock('axios', () => ({
  get: jest.fn(() => Promise.resolve(mockPreferences)),
  patch: jest.fn(() => Promise.reject()),
  create: () => mockAxios,
  defaults: {
    adapter: {},
  },
}));

function factory() {
  return mount(Preferences, {
    localVue,
    router,
  });
}

describe('Preferences.vue', () => {
  describe('Elements are rendered correctly', () => {
    it('Component renders', async () => {
      const wrapper = factory();
      await wrapper.vm.getPreferences();
      wrapper.vm.$nextTick(() => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    it('loading animation is shown', () => {
      const wrapper = factory();
      expect(wrapper.find('.loading-overlay').exists()).toBe(true);
    });

    it('title is rendered', async () => {
      const wrapper = factory();
      await wrapper.vm.getPreferences();
      wrapper.vm.$nextTick(() => {
        expect(wrapper.find('.title').text()).toBe('Preferences for Gunter');
      });
    });
  });

  describe('functionality works', () => {
    it('preferences are saved', async () => {
      const wrapper = factory();
      await wrapper.vm.getPreferences();
      wrapper.vm.$nextTick(() => {
        const spy = jest.spyOn(wrapper.vm, 'savePreferences');
        wrapper.find('.button.is-success').trigger('click');
        expect(spy).toHaveBeenCalled();
      });
    });

    it('notifications can be toggled on', async () => {
      const wrapper = factory();
      await wrapper.vm.getPreferences();
      wrapper.vm.$nextTick(() => {
        wrapper.find('input[type=checkbox]').trigger('click');
        expect(wrapper.vm.notificationsEnabled).toBe(false); // TODO: should be true?
      });
    });

    it('notifications can be toggled off', async () => {
      const wrapper = factory();
      await wrapper.vm.getPreferences();
      wrapper.vm.$nextTick(() => {
        wrapper.find('input[type=checkbox]').trigger('click');
        wrapper.find('input[type=checkbox]').trigger('click');
        expect(wrapper.vm.notificationsEnabled).toBe(true); // TODO: should be false??
      });
    });
  });
});
