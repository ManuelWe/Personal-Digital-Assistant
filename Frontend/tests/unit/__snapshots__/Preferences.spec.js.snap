// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preferences.vue Elements are rendered correctly Component renders 1`] = `
<div id="preferences" class="hero-body">
  <div class="loading-overlay is-active fade-leave fade-leave-active">
    <div class="loading-background"></div>
    <div class="loading-icon"></div>
  </div>
  <!---->
  <div class="container">
    <h1 class="title">
      Preferences for Gunter
    </h1>
    <div class="columns">
      <div class="column">
        <div class="field"><label class="label has-text-white">Name</label>
          <div class="control"><input type="text" placeholder="Name" class="input"></div>
        </div>
        <div class="field"><label class="label has-text-white">Preparation Time in Minutes</label>
          <div class="control"><input type="text" placeholder="Preparation Time in Minutes" class="input"></div>
        </div>
        <div class="field"><label class="label has-text-white">Favourite Quote Category</label>
          <div class="control"><input type="text" placeholder="Favourite Quote Category" class="input"></div>
        </div>
        <div class="field"><label class="label has-text-white">Main City</label>
          <div class="control"><input type="text" placeholder="Main City" class="input"></div>
        </div>
      </div>
      <div class="column">
        <div class="field"><label class="label has-text-white">Current Location</label>
          <div class="control"><input disabled="disabled" type="text" placeholder="Current Location" class="input"></div>
        </div>
        <div class="field"><label class="b-checkbox checkbox"><input type="checkbox" true-value="true" value="false"> <span class="check"></span> <span class="control-label">
            Send notifications to this device
          </span></label></div>
        <div class="field is-grouped is-grouped-right">
          <p class="control"><a href="#/" class="button is-light router-link-exact-active router-link-active">
              Cancel
            </a></p>
          <p class="control"><a class="button is-success">
              Save
            </a></p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
